From 16eb76277f87bae53af288048c6dbc696e96d288 Mon Sep 17 00:00:00 2001
From: StanleyYP Wang <StanleyYP.Wang@mediatek.com>
Date: Mon, 22 Dec 2025 17:03:57 +0800
Subject: [PATCH 017/254] mtk: hostapd: Add three wire PTA ctrl hostapd vendor
 command

Add CEB support
Add runtime cli command
[Usage]
hostapd_cli -i <interface> set_three_wire ext_sel=<val>
ceb_band_sel=<val>

Signed-off-by: StanleyYP Wang <StanleyYP.Wang@mediatek.com>
---
 hostapd/config_file.c             |  8 ++++
 hostapd/ctrl_iface.c              | 44 ++++++++++++++++++++
 hostapd/hostapd_cli.c             |  8 ++++
 src/ap/ap_config.c                |  2 +
 src/ap/ap_config.h                | 26 ++++++++++++
 src/ap/ap_drv_ops.c               | 68 +++++++++++++++++++++++++++++++
 src/ap/ap_drv_ops.h               |  1 +
 src/ap/hostapd.c                  |  2 +
 src/common/mtk_vendor.h           | 13 ++++++
 src/drivers/driver.h              |  9 ++++
 src/drivers/driver_nl80211.c      | 34 ++++++++++++++++
 src/drivers/driver_nl80211.h      |  1 +
 src/drivers/driver_nl80211_capa.c |  3 ++
 13 files changed, 219 insertions(+)

diff --git a/hostapd/config_file.c b/hostapd/config_file.c
index bf6bd3f45..a22e5ae8b 100644
--- a/hostapd/config_file.c
+++ b/hostapd/config_file.c
@@ -5297,6 +5297,14 @@ static int hostapd_config_fill(struct hostapd_config *conf,
 			return 1;
 		}
 		conf->edcca_compensation = (s8) val;
+	} else if (os_strcmp(buf, "three_wire_ext_sel") == 0) {
+		u8 en = atoi(pos);
+
+		conf->three_wire.ext_sel = en;
+	} else if (os_strcmp(buf, "ceb_band_sel") == 0) {
+		u8 en = atoi(pos);
+
+		conf->three_wire.ceb_band_sel = en;
 	} else {
 		wpa_printf(MSG_ERROR,
 			   "Line %d: unknown configuration item '%s'",
diff --git a/hostapd/ctrl_iface.c b/hostapd/ctrl_iface.c
index 9dac985b0..dc4fa653a 100644
--- a/hostapd/ctrl_iface.c
+++ b/hostapd/ctrl_iface.c
@@ -4328,6 +4328,48 @@ hostapd_ctrl_iface_get_mu(struct hostapd_data *hapd, char *buf,
 	}
 }
 
+static int
+hostapd_ctrl_iface_three_wire(struct hostapd_data *hapd, char *cmd,
+			      char *buf, size_t buflen)
+{
+	char *token, *context = NULL;
+	int i, ret;
+
+	while ((token = str_token(cmd, " ", &context))) {
+		if (os_strncmp(token, "ext_sel=", 8) == 0) {
+			hapd->iconf->three_wire.ext_sel = strtol(token + 8, NULL, 10);
+			continue;
+		}
+
+		if (os_strncmp(token, "ceb_band_sel=", 13) == 0) {
+			hapd->iconf->three_wire.ceb_band_sel = strtol(token + 13, NULL, 10);
+			continue;
+		}
+
+		wpa_printf(MSG_ERROR, "CTRL: Invalid THREE_WIRE parameter: %s", token);
+		goto fail;
+	}
+
+	for (i = 0; i < hapd->iface->interfaces->count; i++) {
+		struct hostapd_iface *iface =
+			hapd->iface->interfaces->iface[i];
+
+		if (!iface || iface == hapd->iface)
+			continue;
+
+		iface->conf->three_wire.ext_sel = hapd->iconf->three_wire.ext_sel;
+		iface->conf->three_wire.ceb_band_sel = hapd->iconf->three_wire.ceb_band_sel;
+	}
+
+	ret = hostapd_drv_three_wire_ctrl(hapd);
+	if (ret)
+		goto fail;
+
+	return os_snprintf(buf, buflen, "OK\n");
+
+fail:
+	return os_snprintf(buf, buflen, "FAIL\n");
+}
 
 static int hostapd_ctrl_iface_receive_process(struct hostapd_data *hapd,
 					      char *buf, char *reply,
@@ -4953,6 +4995,8 @@ static int hostapd_ctrl_iface_receive_process(struct hostapd_data *hapd,
 							  reply_size);
 	} else if (os_strncmp(buf, "GET_MU", 6) == 0) {
 		reply_len = hostapd_ctrl_iface_get_mu(hapd, reply, reply_size);
+	} else if (os_strncmp(buf, "SET_THREE_WIRE ", 15) == 0) {
+		reply_len = hostapd_ctrl_iface_three_wire(hapd, buf + 15, reply, reply_size);
 	} else {
 		os_memcpy(reply, "UNKNOWN COMMAND\n", 16);
 		reply_len = 16;
diff --git a/hostapd/hostapd_cli.c b/hostapd/hostapd_cli.c
index 905022fa8..10a175e18 100644
--- a/hostapd/hostapd_cli.c
+++ b/hostapd/hostapd_cli.c
@@ -1472,6 +1472,12 @@ static int hostapd_cli_cmd_get_mu(struct wpa_ctrl *ctrl, int argc,
 	return hostapd_cli_cmd(ctrl, "GET_MU", 0, argc, argv);
 }
 
+static int hostapd_cli_cmd_set_three_wire_ctrl(struct wpa_ctrl *ctrl, int argc,
+					       char *argv[])
+{
+	return hostapd_cli_cmd(ctrl, "SET_THREE_WIRE", 1, argc, argv);
+}
+
 
 #ifdef CONFIG_DPP
 
@@ -1848,6 +1854,8 @@ static const struct hostapd_cli_cmd hostapd_cli_commands[] = {
 		"<value> [0-15] bitmap- UL MU-MIMO(bit3), DL MU-MIMO(bit2), UL OFDMA(bit1), DL OFDMA(bit0)"},
 	{ "get_mu", hostapd_cli_cmd_get_mu, NULL,
 		" = show mu onoff value in 0-15 bitmap"},
+	{ "set_three_wire", hostapd_cli_cmd_set_three_wire_ctrl, NULL,
+		" = Set three wire ctrl" },
 #ifdef CONFIG_DPP
 	{ "dpp_qr_code", hostapd_cli_cmd_dpp_qr_code, NULL,
 	  "report a scanned DPP URI from a QR Code" },
diff --git a/src/ap/ap_config.c b/src/ap/ap_config.c
index af51ab64c..a3b3e257e 100644
--- a/src/ap/ap_config.c
+++ b/src/ap/ap_config.c
@@ -315,6 +315,8 @@ struct hostapd_config * hostapd_config_defaults(void)
 
 	conf->edcca_enable = EDCCA_MODE_AUTO;
 	conf->edcca_compensation = EDCCA_DEFAULT_COMPENSATION;
+	conf->three_wire.ext_sel = THREE_WIRE_EXT_SEL_UNSET;
+	conf->three_wire.ceb_band_sel = THREE_WIRE_CEB_SEL_UNSET;
 
 	hostapd_set_and_check_bw320_offset(conf, 0);
 
diff --git a/src/ap/ap_config.h b/src/ap/ap_config.h
index 957995eda..3f112d6db 100644
--- a/src/ap/ap_config.h
+++ b/src/ap/ap_config.h
@@ -1336,6 +1336,32 @@ struct hostapd_config {
 	u8 edcca_enable;
 	s8 edcca_compensation;
 	int *edcca_threshold;
+	struct {
+		u8 ext_sel;
+		u8 ceb_band_sel;
+	} three_wire;
+};
+
+#define THREE_WIRE_EXT_SEL_UNSET	0xff
+#define THREE_WIRE_EXT_SEL_DISABLE	0
+#define THREE_WIRE_EXT0_ENABLE		BIT(0)
+#define THREE_WIRE_EXT1_ENABLE		BIT(1)
+#define THREE_WIRE_EXT2_ENABLE		BIT(2)
+#define THREE_WIRE_EXT_SEL_MAX		(THREE_WIRE_EXT0_ENABLE | \
+					 THREE_WIRE_EXT1_ENABLE | \
+					 THREE_WIRE_EXT2_ENABLE)
+
+enum three_wire_ceb_band_sel {
+	THREE_WIRE_CEB_SEL_UNSET = 0xff,
+	THREE_WIRE_CEB_SEL_NONE = 0,
+	THREE_WIRE_CEB_SEL_BAND0,
+	THREE_WIRE_CEB_SEL_BAND1,
+	THREE_WIRE_CEB_SEL_BAND2,
+
+	/* keep last */
+	NUM_THREE_WIRE_CEB_BAND_SEL,
+	THREE_WIRE_CEB_BAND_SEL_MAX =
+		NUM_THREE_WIRE_CEB_BAND_SEL - 1
 };
 
 enum edcca_mode {
diff --git a/src/ap/ap_drv_ops.c b/src/ap/ap_drv_ops.c
index 6b542930e..fdcd6b498 100644
--- a/src/ap/ap_drv_ops.c
+++ b/src/ap/ap_drv_ops.c
@@ -1480,3 +1480,71 @@ int hostapd_drv_mu_dump(struct hostapd_data *hapd, u8 *mu_onoff)
 		return 0;
 	return hapd->driver->mu_dump(hapd->drv_priv, mu_onoff);
 }
+
+static void hostapd_update_three_wire_ctrl(struct hostapd_data *hapd)
+{
+	u8 ext_sel = THREE_WIRE_EXT_SEL_UNSET;
+	u8 ceb_band_sel = THREE_WIRE_CEB_SEL_UNSET;
+	int i;
+
+	for (i = 0; i < hapd->iface->interfaces->count; i++) {
+		struct hostapd_iface *iface =
+			hapd->iface->interfaces->iface[i];
+
+		if (!iface)
+			continue;
+
+		if (iface->conf->three_wire.ext_sel != THREE_WIRE_EXT_SEL_UNSET)
+			ext_sel = iface->conf->three_wire.ext_sel;
+		if (iface->conf->three_wire.ceb_band_sel != THREE_WIRE_CEB_SEL_UNSET)
+			ceb_band_sel = iface->conf->three_wire.ceb_band_sel;
+	}
+
+	if (ext_sel == THREE_WIRE_EXT_SEL_UNSET &&
+	    ceb_band_sel == THREE_WIRE_CEB_SEL_UNSET)
+		return;
+
+	for (i = 0; i < hapd->iface->interfaces->count; i++) {
+		struct hostapd_iface *iface =
+			hapd->iface->interfaces->iface[i];
+
+		if (!iface)
+			continue;
+
+		iface->conf->three_wire.ext_sel = ext_sel;
+		iface->conf->three_wire.ceb_band_sel = ceb_band_sel;
+	}
+}
+
+int hostapd_drv_three_wire_ctrl(struct hostapd_data *hapd)
+{
+	u8 ext_sel, ceb_band_sel;
+
+	if (!hapd->driver || !hapd->driver->three_wire_ctrl)
+		return 0;
+
+	if (hapd->iconf->three_wire.ext_sel == THREE_WIRE_EXT_SEL_UNSET &&
+	    hapd->iconf->three_wire.ceb_band_sel == THREE_WIRE_CEB_SEL_UNSET) {
+		hostapd_update_three_wire_ctrl(hapd);
+		return 0;
+	}
+
+	if (hapd->iconf->three_wire.ext_sel != THREE_WIRE_EXT_SEL_UNSET)
+		ext_sel = hapd->iconf->three_wire.ext_sel;
+	else
+		ext_sel = THREE_WIRE_EXT_SEL_DISABLE;
+
+	if (hapd->iconf->three_wire.ceb_band_sel != THREE_WIRE_CEB_SEL_UNSET)
+		ceb_band_sel = hapd->iconf->three_wire.ceb_band_sel;
+	else
+		ceb_band_sel = THREE_WIRE_CEB_SEL_NONE;
+
+	if (ext_sel > THREE_WIRE_EXT_SEL_MAX ||
+	    ceb_band_sel > THREE_WIRE_CEB_BAND_SEL_MAX) {
+		wpa_printf(MSG_ERROR, "Invalid value for three wire control\n");
+		return -1;
+	}
+
+	return hapd->driver->three_wire_ctrl(hapd->drv_priv,
+					     ext_sel, ceb_band_sel);
+}
diff --git a/src/ap/ap_drv_ops.h b/src/ap/ap_drv_ops.h
index a06a8a66b..734f1b79c 100644
--- a/src/ap/ap_drv_ops.h
+++ b/src/ap/ap_drv_ops.h
@@ -164,6 +164,7 @@ int hostapd_drv_configure_edcca_threshold(struct hostapd_data *hapd,
 int hostapd_drv_get_edcca(struct hostapd_data *hapd, const u8 mode, u8 *value);
 int hostapd_drv_mu_ctrl(struct hostapd_data *hapd);
 int hostapd_drv_mu_dump(struct hostapd_data *hapd, u8 *mu_onoff);
+int hostapd_drv_three_wire_ctrl(struct hostapd_data *hapd);
 
 #include "drivers/driver.h"
 
diff --git a/src/ap/hostapd.c b/src/ap/hostapd.c
index 77f97b4f4..5928fac7b 100644
--- a/src/ap/hostapd.c
+++ b/src/ap/hostapd.c
@@ -2825,6 +2825,8 @@ dfs_offload:
 		goto fail;
 	if (hostapd_drv_mu_ctrl(hapd) < 0)
 		goto fail;
+	if (hostapd_drv_three_wire_ctrl(hapd) < 0)
+		goto fail;
 
 	wpa_printf(MSG_DEBUG, "%s: Setup of interface done.",
 		   iface->bss[0]->conf->iface);
diff --git a/src/common/mtk_vendor.h b/src/common/mtk_vendor.h
index 60bc4cd4c..aab662599 100644
--- a/src/common/mtk_vendor.h
+++ b/src/common/mtk_vendor.h
@@ -13,6 +13,7 @@ enum mtk_nl80211_vendor_subcmds {
 	MTK_NL80211_VENDOR_SUBCMD_MU_CTRL = 0xc5,
 	MTK_NL80211_VENDOR_SUBCMD_PHY_CAPA_CTRL= 0xc6,
 	MTK_NL80211_VENDOR_SUBCMD_EDCCA_CTRL = 0xc7,
+	MTK_NL80211_VENDOR_SUBCMD_3WIRE_CTRL = 0xc8
 };
 
 enum mtk_vendor_attr_edcca_ctrl {
@@ -58,6 +59,18 @@ static struct nla_policy edcca_ctrl_policy[NUM_MTK_VENDOR_ATTRS_EDCCA_CTRL] = {
 	[MTK_VENDOR_ATTR_EDCCA_CTRL_SEC160_VAL] = { .type = NLA_U8 },
 };
 
+enum mtk_vendor_attr_3wire_ctrl {
+	MTK_VENDOR_ATTR_3WIRE_CTRL_UNSPEC,
+
+	MTK_VENDOR_ATTR_3WIRE_EXT_SEL,
+	MTK_VENDOR_ATTR_3WIRE_CEB_BAND_SEL,
+
+	/* keep last */
+	NUM_MTK_VENDOR_ATTRS_3WIRE_CTRL,
+	MTK_VENDOR_ATTR_3WIRE_CTRL_MAX =
+		NUM_MTK_VENDOR_ATTRS_3WIRE_CTRL - 1
+};
+
 enum mtk_vendor_attr_csi_ctrl {
 	MTK_VENDOR_ATTR_CSI_CTRL_UNSPEC,
 
diff --git a/src/drivers/driver.h b/src/drivers/driver.h
index 9c3e05423..532809253 100644
--- a/src/drivers/driver.h
+++ b/src/drivers/driver.h
@@ -5560,6 +5560,15 @@ struct wpa_driver_ops {
 	 */
 	 int (*mu_ctrl)(void *priv, u8 mu_onoff);
 	 int (*mu_dump)(void *priv, u8 *mu_onoff);
+
+	/**
+	 * three_wire_ctrl - set three_wire_ctrl mode
+	 * @priv: Private driver interface data
+	 * @ext_sel: external PTA selection for three wire control
+	 * @ceb_band_sel: CEB band selection for three wire control
+	 *
+	 */
+	 int (*three_wire_ctrl)(void *priv, const u8 ext_sel, const u8 ceb_band_sel);
 };
 
 /**
diff --git a/src/drivers/driver_nl80211.c b/src/drivers/driver_nl80211.c
index c910bda9f..56e3b9d19 100644
--- a/src/drivers/driver_nl80211.c
+++ b/src/drivers/driver_nl80211.c
@@ -15333,6 +15333,39 @@ static int nl80211_get_edcca(void *priv, const u8 mode, u8 *value)
 	return ret;
 }
 
+static int nl80211_enable_three_wire(void *priv, const u8 ext_sel, const u8 ceb_band_sel)
+{
+	struct i802_bss *bss = priv;
+	struct wpa_driver_nl80211_data *drv = bss->drv;
+	struct nl_msg *msg;
+	struct nlattr *data;
+	int ret;
+
+	if (!drv->mtk_3wire_vendor_cmd_avail) {
+		wpa_printf(MSG_INFO,
+			   "nl80211: Driver does not support setting three wire control");
+		return 0;
+	}
+
+	if (!(msg = nl80211_drv_msg(drv, 0, NL80211_CMD_VENDOR)) ||
+	    nla_put_u32(msg, NL80211_ATTR_VENDOR_ID, OUI_MTK) ||
+	    nla_put_u32(msg, NL80211_ATTR_VENDOR_SUBCMD,
+			MTK_NL80211_VENDOR_SUBCMD_3WIRE_CTRL) ||
+	    !(data = nla_nest_start(msg, NL80211_ATTR_VENDOR_DATA)) ||
+	    nla_put_u8(msg, MTK_VENDOR_ATTR_3WIRE_EXT_SEL, ext_sel) ||
+	    nla_put_u8(msg, MTK_VENDOR_ATTR_3WIRE_CEB_BAND_SEL, ceb_band_sel)) {
+		nlmsg_free(msg);
+		return -ENOBUFS;
+	}
+	nla_nest_end(msg, data);
+
+	ret = send_and_recv_cmd(drv, msg);
+	if (ret)
+		wpa_printf(MSG_ERROR, "Failed to enable three wire. ret=%d (%s) ",
+			   ret, strerror(-ret));
+
+	return ret;
+}
 
 static struct hostapd_multi_hw_info *
 wpa_driver_get_multi_hw_info(void *priv, unsigned int *num_multi_hws)
@@ -15522,4 +15555,5 @@ const struct wpa_driver_ops wpa_driver_nl80211_ops = {
 	.configure_edcca_enable = nl80211_configure_edcca_enable,
 	.configure_edcca_threshold = nl80211_configure_edcca_threshold,
 	.get_edcca = nl80211_get_edcca,
+	.three_wire_ctrl = nl80211_enable_three_wire,
 };
diff --git a/src/drivers/driver_nl80211.h b/src/drivers/driver_nl80211.h
index 6c787b630..78bd3950b 100644
--- a/src/drivers/driver_nl80211.h
+++ b/src/drivers/driver_nl80211.h
@@ -203,6 +203,7 @@ struct wpa_driver_nl80211_data {
 	unsigned int connect_ext_vendor_cmd_avail:1;
 	unsigned int mtk_edcca_vendor_cmd_avail:1;
 	unsigned int mtk_mu_vendor_cmd_avail:1;
+	unsigned int mtk_3wire_vendor_cmd_avail:1;
 
 	u8 extra_bss_membership_selectors[8];
 
diff --git a/src/drivers/driver_nl80211_capa.c b/src/drivers/driver_nl80211_capa.c
index a525a9afd..56c8a9f60 100644
--- a/src/drivers/driver_nl80211_capa.c
+++ b/src/drivers/driver_nl80211_capa.c
@@ -1155,6 +1155,9 @@ static int wiphy_info_handler(struct nl_msg *msg, void *arg)
 				case MTK_NL80211_VENDOR_SUBCMD_MU_CTRL :
 					drv->mtk_mu_vendor_cmd_avail = 1;
 					break;
+				case MTK_NL80211_VENDOR_SUBCMD_3WIRE_CTRL :
+					drv->mtk_3wire_vendor_cmd_avail = 1;
+					break;
 				}
 			}
 
-- 
2.45.2

